File: src/config/bootstrap.ts
import { access, copyFile, cp, mkdir, rm, writeFile } from "node:fs/promises";
import { constants } from "node:fs";
import { dirname } from "node:path";
import type { ConfigPaths } from "./paths.js";

export type BootstrapMode = "refresh" | "repair";

/** Paths under ~/.qc/ that packaged bootstrap writes into .gitignore. */
export const REFERENCE_GITIGNORE_ENTRIES = [".default-settings/", "sessions/"] as const;

/** Repair path used by normal qc invocations; refresh is for postinstall/install-current. */
export async function bootstrapGlobal(paths: ConfigPaths): Promise<void> {
  await bootstrapSettings("repair", paths);
}

async function pathExists(path: string): Promise<boolean> {
  try {
    await access(path, constants.F_OK);
    return true;
  } catch (cause: unknown) {
    if ((cause as NodeJS.ErrnoException).code === "ENOENT") return false;
    throw cause;
  }
}

/** Create ~/.qc through an existing symlink without replacing that inode. */
async function ensureGlobalRoot(paths: ConfigPaths): Promise<void> {
  await mkdir(paths.globalRoot, { recursive: true });
}

/** Force-replace the packaged settings reference mirror under ~/.qc/.default-settings/. */
async function refreshDefaultSettings(paths: ConfigPaths): Promise<void> {
  await ensureGlobalRoot(paths);
  await rm(paths.defaultSettings, { recursive: true, force: true });
  await cp(paths.bundledSettings, paths.defaultSettings, { recursive: true, force: true });
}

/** Overwrite ~/.qc/.gitignore with package-owned ignore entries. */
async function refreshGitignore(paths: ConfigPaths): Promise<void> {
  await ensureGlobalRoot(paths);
  // Session mappings are runtime state; keep them out of a versioned ~/.qc/.
  await writeFile(paths.gitignore, `${REFERENCE_GITIGNORE_ENTRIES.join("\n")}\n`);
}

/** Create ~/.qc/.gitignore only when absent (repair mode). */
async function ensureGitignore(paths: ConfigPaths): Promise<void> {
  if (await pathExists(paths.gitignore)) return;
  await refreshGitignore(paths);
}

/** Copy packaged config.toml when the bootstrap sentinel is missing. Never overwrites. */
async function seedConfig(paths: ConfigPaths): Promise<void> {
  await ensureGlobalRoot(paths);
  await mkdir(dirname(paths.global), { recursive: true });
  try {
    await copyFile(paths.starter, paths.global, constants.COPYFILE_EXCL);
  } catch (cause: unknown) {
    if (!(cause instanceof Error) || (cause as NodeJS.ErrnoException).code !== "EEXIST") throw cause;
  }
}

/** Wipe and recopy packaged sample prompts into ~/.qc/prompts/samples/. */
export async function hardRefreshSamplePrompts(paths: ConfigPaths): Promise<void> {
  await ensureGlobalRoot(paths);
  await mkdir(dirname(paths.samplePrompts), { recursive: true });
  await rm(paths.samplePrompts, { recursive: true, force: true });
  await cp(paths.starterSamplePrompts, paths.samplePrompts, { recursive: true, force: true });
}

/** CLI flag: refresh samples only; does not create config.toml. */
export async function installSamplePrompts(paths: ConfigPaths): Promise<void> {
  await hardRefreshSamplePrompts(paths);
}

export async function bootstrapSettings(mode: BootstrapMode, paths: ConfigPaths): Promise<void> {
  if (mode === "refresh") {
    await refreshDefaultSettings(paths);
    await refreshGitignore(paths);
  } else if (!(await pathExists(paths.defaultSettings))) {
    await refreshDefaultSettings(paths);
    await ensureGitignore(paths);
  } else {
    await ensureGitignore(paths);
  }

  const configMissing = !(await pathExists(paths.global));
  if (configMissing) {
    await seedConfig(paths);
    await hardRefreshSamplePrompts(paths);
  }
}

File: src/config/bootstrap.rs
use super::paths::ConfigPaths;
use crate::errors::QcError;
use crate::fsutil::{copy_tree, ensure_dir, path_exists, qc_io, remove_tree};
use std::fs;
use std::io::{self, Write};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BootstrapMode {
    Refresh,
    Repair,
}

/// Paths under ~/.qc/ that packaged bootstrap writes into .gitignore.
pub const REFERENCE_GITIGNORE_ENTRIES: [&str; 2] = [".default-settings/", "sessions/"];

pub fn bootstrap_global(paths: &ConfigPaths) -> Result<(), QcError> {
    bootstrap_settings(BootstrapMode::Repair, paths)
}

fn refresh_default_settings(paths: &ConfigPaths) -> Result<(), QcError> {
    ensure_dir(&paths.global_root)?;
    remove_tree(&paths.default_settings)?;
    copy_tree(&paths.bundled_settings, &paths.default_settings)
}

fn refresh_gitignore(paths: &ConfigPaths) -> Result<(), QcError> {
    ensure_dir(&paths.global_root)?;
    let body = format!("{}\n", REFERENCE_GITIGNORE_ENTRIES.join("\n"));
    fs::write(&paths.gitignore, body).map_err(|cause| qc_io("write", &paths.gitignore, cause))
}

fn ensure_gitignore(paths: &ConfigPaths) -> Result<(), QcError> {
    if path_exists(&paths.gitignore)? {
        return Ok(());
    }
    refresh_gitignore(paths)
}

/// Create config.toml only when absent. Never overwrite.
fn seed_config(paths: &ConfigPaths) -> Result<(), QcError> {
    ensure_dir(&paths.global_root)?;
    if let Some(parent) = paths.global.parent() {
        ensure_dir(parent)?;
    }
    let mut file = match fs::OpenOptions::new()
        .write(true)
        .create_new(true)
        .open(&paths.global)
    {
        Ok(file) => file,
        Err(err) if err.kind() == io::ErrorKind::AlreadyExists => return Ok(()),
        Err(err) => return Err(qc_io("create", &paths.global, err)),
    };
    let source = fs::read(&paths.starter).map_err(|cause| qc_io("read", &paths.starter, cause))?;
    file.write_all(&source)
        .map_err(|cause| qc_io("write", &paths.global, cause))?;
    Ok(())
}

/// Wipe and recopy packaged sample prompts into ~/.qc/prompts/samples/.
pub fn hard_refresh_sample_prompts(paths: &ConfigPaths) -> Result<(), QcError> {
    ensure_dir(&paths.global_root)?;
    if let Some(parent) = paths.sample_prompts.parent() {
        ensure_dir(parent)?;
    }
    remove_tree(&paths.sample_prompts)?;
    copy_tree(&paths.starter_sample_prompts, &paths.sample_prompts)
}

/// CLI flag: refresh samples only; does not create config.toml.
pub fn install_sample_prompts(paths: &ConfigPaths) -> Result<(), QcError> {
    hard_refresh_sample_prompts(paths)
}

pub fn bootstrap_settings(mode: BootstrapMode, paths: &ConfigPaths) -> Result<(), QcError> {
    if mode == BootstrapMode::Refresh {
        refresh_default_settings(paths)?;
        refresh_gitignore(paths)?;
    } else if !path_exists(&paths.default_settings)? {
        refresh_default_settings(paths)?;
        ensure_gitignore(paths)?;
    } else {
        ensure_gitignore(paths)?;
    }

    if !path_exists(&paths.global)? {
        seed_config(paths)?;
        hard_refresh_sample_prompts(paths)?;
    }
    Ok(())
}
